---
title: "run_seurat_preprocessig"
output: html_document
date: "2025-06-26"
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#knitr::opts_knit$set(root.dir = "/media/labct/P4/Single_cell_rna_seq_analyses_eline/sc3samples_whitelist")
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```


```{r}
rootdir="/media/inserm-root/P4/Single_cell_rna_seq_analyses_eline/sc3samples_whitelist"
#setwd(rootdir)  # Set this to correct location

#source("/media/inserm-root/P4/Single_cell_rna_seq_analyses_eline/project/R/seurat_processing.R")
library(SingleCellExperiment)
library(FLAMES)
library(rtracklayer)
library(tidyverse)
```

(super long):
```{r}
sample_list=c("ctrl","S24","R")

cov_list<-lapply(sample_list,function(x) {
  bam=paste0(rootdir,"/flames/",x,"/realign2transcript.bam")
  p1<-plot_coverage(bam)
  p1
  ggsave(filename = paste0(rootdir,"/flames/coverage_plot_",x,".png"),plot = p1)
  return(p1)
  
})
cov_list
names(cov_list)<-sample_list
print(cov_list)
#saveRDS(cov_list, file = paste0(rootdir,"/flames/coverage_plots.rds"))
```


```{r eval=FALSE}


sample_list=c("ctrl","S24","R")

cov_list_df<-lapply(sample_list,function(x) {
  bam=paste0(rootdir,"/flames/",x,"/realign2transcript.bam")
  cov<-get_coverage(bam, min_counts = 2, remove_UTR = FALSE)

  return(cov)
  
})
cov_list_df
names(cov_list)<-sample_list
```

other coverage plots:
```{r fig.width=10}

lapply(sample_list,function(x) {
  #read flames' gff file
  # x<-"ctrl"
  gff <- import(paste0(rootdir,"/flames/",x,"/isoform_annotated.gff3"))
  # only keep transcripts
  transcripts <- gff[!is.na(gff$transcript_id)]
  
  df<-as.data.frame(transcripts)
  df<-df[,c("seqnames","support_count","transcript_id")]
  head(df)
  df <- df |> mutate(support_count = as.numeric(support_count))

  chrom<-df|>group_by(seqnames) |> summarise(count_per_chrom=sum(!is.na(support_count)))



  chrom_mean<-df|>group_by(seqnames) |> summarise(mean_support_count = mean(support_count, na.rm = TRUE))
  chrom_mean <- chrom_mean |>
  mutate(mean_support_count_rounded = round(mean_support_count))
  chrom_mean

  # plot support read mean count for each chromososme
plotmeanchrom<-ggplot(chrom_mean, aes(x = seqnames, y = mean_support_count)) +
  geom_bar(stat = "identity", fill = "lightblue"
           ) +
  geom_text(
    aes(label = mean_support_count_rounded),
    vjust = -0.5, size = 5
  ) +
  labs(title = paste0(x," sample, support read per transcript mean count for each chromososme"),
       x = "Chromosome",
       y = "Support counts per transcript mean") +
  theme_minimal() 
print(plotmeanchrom)
ggsave(filename = paste0(rootdir,"/flames/coverage_plots_chrom_",x,".png"),plotmeanchrom)

# plot support_count (coverage) distribution across transcripts
plot_coverage<-ggplot(df, aes(x = support_count)) +
   geom_histogram(binwidth = 5, fill = "lightgreen", color = "darkgreen") +
  geom_text(stat = "bin",
             binwidth = 5,
             aes(label = ..count.., y = ..count..),
             vjust = -1, size = 3,angle = 45,nudge_x=2) +
   scale_x_continuous(limits = c(0, 500)) +
   labs(
     title = paste0(x, " sample, support_count (coverage) distribution across transcripts"),
     x = "Support Count (Coverage)",
     y = "Number of transcripts"
   ) +
   theme_minimal()

print(plot_coverage)
ggsave(filename = paste0(rootdir,"/flames/coverage_plots_transcriptome_",x,".png"),plot_coverage)

plot_coverage2<-ggplot(df, aes(x = transcript_id, y = support_count)) +
     geom_bar(stat = "identity",  fill = "lightgreen", color = "darkgreen") +
     labs(
         title = paste0(x, " sample, support_count per transcript"),
         x = "Transcript",
        y = "Support Count (Coverage)"
     ) +
     theme_minimal() +
     theme(
         axis.text.x = element_blank(),   # Cache les labels X si trop nombreux
         axis.ticks.x = element_blank()
     )
print(plot_coverage2)
ggsave(filename = paste0(rootdir,"/flames/coverage_plots_transcriptome2_",x,".png"),plot_coverage2)
})
```



```{r eval=FALSE}
x="ctrl"
sce<-readRDS(paste0(rootdir,"/flames/",x,"/sce.rds"))
t<-sc_DTU_analysis(
  sce
)
```

