---
title: "02-Seurat_processing"
output: html_document
---


# QC and creation of Seurat Objects
```{r, message=FALSE}
source("./functions/seurat_processing.R")
dir.create(paste0(rootdir,"/QC"))
```
Load the count matrices from the previous step
```{r}
mat_ctrl_iso  <- read.csv(paste0(rootdir,"/counts/gene_symbol_ctrltranscript_level_counts.csv"), row.names = 1)
mat_ctrl_gene <- read.csv(paste0(rootdir,"/counts/gene_symbol_ctrlgene_level_counts.csv"), row.names = 1)

mat_S24_iso   <- read.csv(paste0(rootdir,"/counts/gene_symbol_S24transcript_level_counts.csv"), row.names = 1)
mat_S24_gene  <- read.csv(paste0(rootdir,"/counts/gene_symbol_S24gene_level_counts.csv"), row.names = 1)

mat_R_iso     <- read.csv(paste0(rootdir,"/counts/gene_symbol_Rtranscript_level_counts.csv"), row.names = 1)
mat_R_gene    <- read.csv(paste0(rootdir,"/counts/gene_symbol_Rgene_level_counts.csv"), row.names = 1)

counts_ctrl <- list(
  iso  = mat_ctrl_iso,
  gene = mat_ctrl_gene
)

counts_S24 <- list(
  iso  = mat_S24_iso,
  gene = mat_S24_gene
)

counts_R <- list(
  iso  = mat_R_iso,
  gene = mat_R_gene
)
```


Run QC on both iso-level and gene-level objects.
This function runs a standard Seurat QC, with adaptable filters on the threshold of minimum and maximum features and UMI per cells, and on the percentage of mitochondrial genome. 

```{r qc, eval=T}
res_ctrl_iso<-perform_Seurat_QC("ctrl",
                                counts_ctrl$iso,
                                type="isoform", #specify if you are working with an isoform- or a gene-level count matrix
                                max.features =2000,
                                min.features = 200,
                                min.counts = 400,
                                max.counts = 5000,
                                MT = 10,
                                npc = 15,
                                output_path=paste0(rootdir,"/QC")
)
res_ctrl_iso$seu_obj
res_ctrl_iso$features_before
res_ctrl_iso$features_after
res_ctrl_iso$mt_before
res_ctrl_iso$mt_after
res_ctrl_iso$scatter
seurat_ctrl_iso<-res_ctrl_iso$seu_obj

res_ctrl_gene<-perform_Seurat_QC("ctrl",
                                 counts_ctrl$gene,
                                 type="gene", #specify if you are working with an isoform- or a gene-level count matrix
                                 max.features =2000,
                                 min.features = 200,
                                 min.counts = 400,
                                 max.counts = 5000,
                                 MT = 10,
                                 npc = 15,
                                 output_path=paste0(rootdir,"/QC")
)
res_ctrl_gene$seu_obj
res_ctrl_gene$features_before
res_ctrl_gene$features_after
res_ctrl_gene$mt_before
res_ctrl_gene$mt_after
res_ctrl_gene$scatter
seurat_ctrl_gene<-res_ctrl_gene$seu_obj



#r S24_seu}
res_S24_iso<-perform_Seurat_QC("S24",
                               counts_S24$iso,
                               type="isoform", #specify if you are working with an isoform- or a gene-level count matrix
                               max.features =2000,
                               min.features = 200,
                               min.counts = 400,
                               max.counts = 5000,
                               MT = 10,
                               npc = 15,
                               output_path=paste0(rootdir,"/QC")
)

res_S24_iso$seu_obj
res_S24_iso$features_before
res_S24_iso$features_after
res_S24_iso$mt_before
res_S24_iso$mt_after
res_S24_iso$scatter
seurat_S24_iso<-res_S24_iso$seu_obj

res_S24_gene<-perform_Seurat_QC("S24",
                                counts_S24$gene,
                                type="gene", #specify if you are working with an isoform- or a gene-level count matrix
                                max.features =2000,
                                min.features = 200,
                                min.counts = 400,
                                max.counts = 5000,
                                MT = 10,
                                npc = 15,
                                output_path=paste0(rootdir,"/QC")
)
res_S24_gene$seu_obj
res_S24_gene$features_before
res_S24_gene$features_after
res_S24_gene$mt_before
res_S24_gene$mt_after
res_S24_gene$scatter
seurat_S24_gene<-res_S24_gene$seu_obj



#r R_seu}
res_R_iso<-perform_Seurat_QC("R",
                             counts_R$iso,
                             type="isoform", #specify if you are working with an isoform- or a gene-level count matrix
                             max.features =2000,
                             min.features = 200,
                             min.counts = 400,
                             max.counts = 5000,
                             MT = 10,
                             npc = 15,
                             output_path=paste0(rootdir,"/QC")
)
res_R_iso$seu_obj
res_R_iso$features_before
res_R_iso$features_after
res_R_iso$mt_before
res_R_iso$mt_after
res_R_iso$scatter
seurat_R_iso<-res_R_iso$seu_obj

res_R_gene<-perform_Seurat_QC("R",
                              counts_R$gene,
                              type="gene", #specify if you are working with an isoform- or a gene-level count matrix
                              max.features =2000,
                              min.features = 200,
                              min.counts = 400,
                              max.counts = 5000,
                              MT = 10,
                              npc = 15,
                              output_path=paste0(rootdir,"/QC")
)

res_R_gene$seu_obj
res_R_gene$features_before
res_R_gene$features_after
res_R_gene$mt_before
res_R_gene$mt_after
res_R_gene$scatter
seurat_R_gene<-res_R_gene$seu_obj
```


# Merge and regression after regression on cell cycle and ribosomal genes
Only for demonstration purposes. The objects created with this function will not be used in further analysis, because the diversity in  ribosomal and histonic genes might hide some more subtle signals. 

## At transcript level

```{r}
dir.create(paste0(rootdir,"/merge_integration_ribo_histo"))
```

First without interation:
```{r merge_iso}
seu_objects<-list(seurat_ctrl_iso,seurat_S24_iso,seurat_R_iso)
sample_names<-list("ctrl","S24","R")

merged_iso<-run_regression_and_merge(sample_names, # a list of the sample's names, in the right order
                                 seu_objects, # a list of seurat objects after QC (same order as their names)
                                 type="isoform",
                                 output_path=paste0(rootdir,"/merge_integration_ribo_histo"),
                               integrateRPCA=FALSE)
```

And with integration:
```{r}
seu_objects<-list(seurat_ctrl_iso,seurat_S24_iso,seurat_R_iso)
sample_names<-list("ctrl","S24","R")

merged_iso<-run_regression_and_merge(sample_names, # a list of the sample's names, in the right order
                                 seu_objects, # a list of seurat objects after QC (same order as their names)
                                 type="isoform",
                                 output_path=paste0(rootdir,"/merge_integration_ribo_histo"),
                               integrateRPCA=TRUE)
```


The global stats on the merged objects: we plot the number of isoforms per gene and the genes with the most isoforms.

```{r glob_stats}
number_of_isoforms_per_gene(merged_iso,"merged", outdir=paste0(rootdir,"/merge_integration_ribo_histo"))
top10_gene_with_most_iso(merged_iso,"merged", outdir=paste0(rootdir,"/merge_integration_ribo_histo"))
```

## At gene level

```{r}
seurat_ctrl_gene<-readRDS(paste0(rootdir,"/QC/seu_obj_ctrl_gene.rds"))
seurat_S24_gene<-readRDS(paste0(rootdir,"/QC/seu_obj_S24_gene.rds"))
seurat_R_gene<-readRDS(paste0(rootdir,"/QC/seu_obj_R_gene.rds"))
```
First without interation:
```{r merge_gene}
seu_objects<-list(seurat_ctrl_gene,seurat_S24_gene,seurat_R_gene)
sample_names<-list("ctrl","S24","R")
merged_gene<-run_regression_and_merge(sample_names, # a list of the sample's names, in the right order
                                  seu_objects, # a list of seurat objects after QC (same order as their names)
                                  type="gene",
                                  output_path=paste0(rootdir,"/merge_integration_ribo_histo"),
                               integrateRPCA=FALSE)


```
And with integration:
```{r}
seu_objects<-list(seurat_ctrl_gene,seurat_S24_gene,seurat_R_gene)
sample_names<-list("ctrl","S24","R")
merged_gene<-run_regression_and_merge(sample_names, # a list of the sample's names, in the right order
                                  seu_objects, # a list of seurat objects after QC (same order as their names)
                                  type="gene",
                                  output_path=paste0(rootdir,"/merge_integration_ribo_histo"),
                               integrateRPCA=TRUE)


```

These venn diagramms shows the number of features common to each sample. 

```{r, eval=FALSE}
plot_venn(merged_gene,"gene", output=paste0(rootdir,"/merge_integration_ribo_histo"))
plot_venn(merged_iso,"isoform")
```
